### PR TITLE
Add "sync_mods" argument to state.apply/state.sls

### LINKED ISSUE #48025: Autoloading Utility Modules?
### Description of Issue/Question
In https://docs.saltstack.com/en/2017.7/topics/utils/index.html it is written that 

> Utility modules placed in salt://_utils/ will be synced to the minions when any of the following Salt functions are called:
>   - state.apply
>   - saltutil.sync_utils
>   - saltutil.sync_all

But following the examples displayed on that page does not seem to work.
`state.apply` does not appear to sync anything from _utils.

### Steps to Reproduce Issue
_utils/foo.py
```
# -*- coding: utf-8 -*-
'''
My utils module
---------------

This module contains common functions for use in my other custom types.
'''

def bar():
    return 'baz'

```

foo.sls state
```
#!py
import salt.config
import salt.loader

# required? else __utils__ not available...
__opts__ = salt.config.minion_config('/etc/salt/minion')
__utils__ = salt.loader.utils(__opts__)


def run():
    print __utils__['foo.bar']()

```
command
```
salt-call state.apply foo

local:
    Data failed to compile:
----------
    Rendering SLS 'base:foo' failed: Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/salt/utils/templates.py", line 589, in py
    data = mod.run()
  File "/var/cache/salt/minion/files/base/foo.sls", line 11, in run
    print __utils__['foo.bar']()
  File "/usr/lib/python2.7/site-packages/salt/loader.py", line 1117, in __getitem__
    func = super(LazyLoader, self).__getitem__(item)
  File "/usr/lib/python2.7/site-packages/salt/utils/lazy.py", line 101, in __getitem__
    raise KeyError(key)
KeyError: 'foo.bar'

```

with saltutil.sync_utils
```
salt-call saltutil.sync_utils; salt-call state.apply foo

local:
    - utils.foo

baz
baz

local:
Summary for local
-----------
Succeeded: 0
Failed:   0
-----------
Total states run:    0
Total run time:  0.000 ms

```

### Versions Report
```
Salt Version:
           Salt: 2017.7.5
 
Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: Not Installed
      docker-py: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.7.2
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.5.1
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 2.7.5 (default, Apr 11 2018, 07:36:10)
   python-gnupg: Not Installed
         PyYAML: 3.11
          PyZMQ: 15.3.0
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.1.4
 
System Versions:
           dist: centos 7.5.1804 Core
         locale: UTF-8
        machine: x86_64
        release: 3.10.0-862.3.2.el7.x86_64
         system: Linux
        version: CentOS Linux 7.5.1804 Core

```

### PR DESCRIPTION
**Pull request by terminalmage**
The docs inaccurately stated that custom module types were synced when `state.apply` is run, but this is only the case when run without SLS targets (i.e. when a highstate is run). This PR does the following:

1. Updates the docs

2. Adds a "sync" argument to provide the functionality that the docs suggested was already present

Resolves #48025.

### CODE REVIEW

**Review by rallytime — APPROVED** [review]

File: salt/modules/state.py [hunk 1]
```diff
@@ -599,6 +598,22 @@ def apply_(mods=None,
         .. code-block:: bash
 
             salt '*' state.apply test localconfig=/path/to/minion.yml
+
+    sync
```
Comment (cachedout) [inline salt/modules/state.py]: Should this be something like `sync_mods`. The first thing this makes me think of is `synchronous` but maybe that's just me?

Comment (rallytime) [reply salt/modules/state.py]: I like that idea ^

Comment (terminalmage) [reply salt/modules/state.py]: Yes, that does sound better, I will get that changed.

**Comment by terminalmage** [comment]
`sync` renamed to `sync_mods` per recommendation from @cachedout.

Also, I updated the versionadded to bump the version number to 2018.3.3.

**Comment by rallytime** [comment]
@terminalmage There is a related test failure with this change:

- unit.modules.test_state.StateTestCase.test_sls_sync

https://jenkins.saltstack.com/job/PR/job/salt-pr-linode-ubuntu14-n/23710/

Can you take a look?

**Comment by terminalmage** [comment]
This is fixed now, I forgot to update my test to reflect the change of the argument name.

File: salt/modules/state.py [hunk 2]
```diff
@@ -599,6 +598,22 @@ def apply_(mods=None,
...
+
+            salt '*' state.apply test sync_mods=states,modules
+            salt '*' state.apply test sync_mods=all
+
+        .. note::
+            This option is ignored when no SLS files are specified, as a
+            :ref:`highstate <running-highstate>` automatically syncs all custom
+            module types.
+
+        .. versionadded:: 2017.7.7,2018.3.3,Fluorine
```
Comment (rallytime) [inline salt/modules/state.py]: Ah, i just noticed this since you mentioned updating 2018.3.3 in another comment. The `2017.7.7` should be `2017.7.8`

Comment (terminalmage) [reply salt/modules/state.py]: Done.

**Pull request merged**